python/rbac/upgrade_role.py

```diff
@@ -0,0 +1,150 @@
+from django.contrib.auth.models import Permission
+
+from adwp_base.errors import raise_AdwpEx as err
+
+import cm.checker
+from rbac import log
+from rbac.models import Role, RoleMigration
+
+
+BASE_DIR = dirname(dirname(dirname(os.path.abspath(__file__))))
+CODE_DIR = os.path.join(BASE_DIR, 'python')
+ROLE_SPEC = os.path.join(CODE_DIR, 'rbac', 'role_spec.yaml')
+ROLE_SCHEMA = os.path.join(CODE_DIR, 'rbac', 'role_schema.yaml')
+
+
+def upgrade(data):
+    new_roles = {}
+    for role in data['roles']:
+        new_roles[role['name']] = upgrade_role(role, data)
+
+    for role in data['roles']:
+        role_obj = new_roles[role['name']]
+        role_obj.childs.clear()
+        if 'childs' not in role:
+            continue
+        for child in role['childs']:
+            child_role = new_roles[child]
+            role_obj.childs.add(child_role)
+        role_obj.save()
+
+    for role in new_roles.values():
+        perm_list = role.get_permissions()
+        for user in role.user.all():
+            user.user_permissions.clear()
+            for perm in perm_list:
+                user.user_permissions.add(perm)
+            user.save()
+
+
+def find_role(name, roles):
+    for role in roles:
+        if role['name'] == name:
+            return role
+    return err('INVALID_ROLE_SPEC', f'child role "{name}" is absent')
+
+
+def check_roles_childs(data):
+    for role in data['roles']:
+        if 'childs' in role:
+            for child in role['childs']:
+                find_role(child, data['roles'])
+
+
+def get_role_permissions(role, data):
+    all_perm = []
+    if 'apps' not in role:
+        return []
+    for app in role['apps']:
+        for model in app['models']:
+            try:
+                ct = ContentType.objects.get(app_label=app['label'], model=model['name'])
+            except ContentType.DoesNotExist:
+                msg = 'no model "{}" in application "{}"'
+                err('INVALID_ROLE_SPEC', msg.format(model['name'], app['label']))
+            for code in model['codenames']:
+                codename = f"{code}_{model['name']}"
+                try:
+                    perm = Permission.objects.get(content_type=ct, codename=codename)
+                except Permission.DoesNotExist:
+                    err('INVALID_ROLE_SPEC', f'permission with codename "{codename}" is not found')
+                if perm not in all_perm:
+                    all_perm.append(perm)
+    return all_perm
+
+
+def upgrade_role(role, data):
+    perm_list = get_role_permissions(role, data['roles'])
+    try:
+        new_role = Role.objects.get(name=role['name'])
+        new_role.permissions.clear()
+    except Role.DoesNotExist:
+        new_role = Role(name=role['name'])
+        new_role.save()
+    if 'description' in role:
+        new_role.description = role['description']
+    for perm in perm_list:
+        new_role.permissions.add(perm)
+    new_role.save()
+    return new_role
+
+
+def get_role_spec():
```

Comment:
I don't like this. This function is not testable for no reason. I think it is better to add parameter "role_filename" or something like this and apply a constant in call time.